File: PromoStudio.Web/Scripts/models/customerVideo.js
﻿/// <reference path="../vsdoc/require.js" />
/// <reference path="../vsdoc/knockout-2.3.0.debug.js" />
/// <reference path="customerVideoVoiceOver.js" />
/// <reference path="customerVideoItem.js" />
/// <reference path="storyboard.js" />

"use strict";

define([
    "models/customerVideoItem",
    "models/storyboard",
    "models/customerVideoVoiceOver",
    "knockout"
], function (
        customerVideoItem,
        storyboard,
        customerVideoVoiceOver,
        ko) {
    function ctor (data) {
        var self = this;
        data = data || {};

        self.pk_CustomerVideoId = ko.observable(data.pk_CustomerVideoId || null);
        self.fk_CustomerId = ko.observable(data.fk_CustomerId || null);
        self.fk_CustomerVideoRenderStatusId = ko.observable(data.fk_CustomerVideoRenderStatusId || 1); // pending
        self.fk_StoryboardId = ko.observable(data.fk_StoryboardId || null);
        self.Name = ko.observable(data.Name || null);
        self.Description = ko.observable(data.Description || null);
        self.RenderFailureMessage = ko.observable(data.RenderFailureMessage || null);
        self.DateCreated = ko.observable(data.DateCreated || null);
        self.DateUpdated = ko.observable(data.DateUpdated || null);
        self.DateCompleted = ko.observable(data.DateCompleted || null);
        self.PreviewFilePath = ko.observable(data.PreviewFilePath || null);
        self.CompletedFilePath = ko.observable(data.CompletedFilePath || null);
        self.VidyardVideoId = ko.observable(data.VidyardVideoId || null);
        self.VidyardPlayerId = ko.observable(data.VidyardPlayerId || null);
        self.VidyardPlayerUuid = ko.observable(data.VidyardPlayerUuid || null);

        self.Storyboard = ko.observable(null);
        self.VoiceOver = ko.observable(null);
        self.Items = ko.observableArray([]);

        self.LoadItems = function (storyboardData, voiceOverData, items) {
            var i, item;
            items = items || [];

            storyboardData = storyboardData || {};
            storyboardData = new storyboard(storyboardData);

            if (voiceOverData) {
                self.VoiceOver(new customerVideoVoiceOver(voiceOverData));
            }

            for (i = 0; i < items.length; i++) {
                item = items[i];
                items[i] = new customerVideoItem(item);
            }

            self.Storyboard(storyboardData);
            self.Items(items);
        };
        self.LoadItems(data.Storyboard, data.VoiceOver, data.Items);
    }

    ctor.prototype.toJSON = function () {
        var copy = ko.toJS(this);
        // remove any unneeded properties
        delete copy.DateCreated;
        delete copy.DateUpdated;
        delete copy.DateCompleted;
        delete copy.RenderFailureMessage;
        delete copy.PreviewFilePath;
        delete copy.CompletedFilePath;
        delete copy.Storyboard;
        delete copy.CustomerVideoRenderStatus;

        return copy;
    };
    return ctor;
});

File: PromoStudio.Common/Models/CustomerVideo.cs
﻿using PromoStudio.Common.Enumerations;
using System;
using System.Collections.Generic;
using System.Linq;

namespace PromoStudio.Common.Models
{
    public class CustomerVideo
    {
        private List<CustomerVideoItem> _items = new List<CustomerVideoItem>();

        public long pk_CustomerVideoId { get; set; }
        public long fk_CustomerId { get; set; }
        public sbyte fk_CustomerVideoRenderStatusId { get; set; }
        public int fk_StoryboardId { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public string RenderFailureMessage { get; set; }
        public DateTime DateCreated { get; set; }
        public DateTime? DateUpdated { get; set; }
        public DateTime? DateCompleted { get; set; }
        public string PreviewFilePath { get; set; }
        public string CompletedFilePath { get; set; }
        public long? VidyardVideoId { get; set; }
        public long? VidyardPlayerId { get; set; }
        public string VidyardPlayerUuid { get; set; }

        public CustomerVideoRenderStatus RenderStatus
        {
            get { return (CustomerVideoRenderStatus)fk_CustomerVideoRenderStatusId; }
            set { fk_CustomerVideoRenderStatusId = (sbyte)value; }
        }

        public Storyboard Storyboard { get; set; }
        public CustomerVideoVoiceOver VoiceOver { get; set; }

        public List<CustomerVideoItem> Items
        {
            get { return _items; }
            set { _items = value; }
        }

        public string GetVideoItemsJson(bool isPreview)
        {
            if (Items == null) { return "[]"; }
            return string.Format("[{0}]", string.Join(",",
                Items
                .Where(i =>
                    i.Type == CustomerVideoItemType.StockVideo
                    || (i.Type == CustomerVideoItemType.CustomerTemplateScript))
                .OrderBy(i => i.SortOrder ?? 0)
                .Select(i => i.GetRenderItemJson(isPreview))
                .Where(json => json != null)));
        }

        public string GetAudioItemsJson()
        {
            if (Items == null) { return "[]"; }
            return string.Format("[{0}]", string.Join(",",
                Items
                .Where(i =>
                    i.Type == CustomerVideoItemType.StockAudio
                    || (i.Type == CustomerVideoItemType.CustomerVideoVoiceOver))
                .Select(i => i.GetRenderItemJson(false))
                .Where(json => json != null)));
        }

        public dynamic ToPoco()
        {
            return new
            {
                pk_CustomerVideoId = pk_CustomerVideoId,
                fk_CustomerId = fk_CustomerId,
                fk_CustomerVideoRenderStatusId = fk_CustomerVideoRenderStatusId,
                fk_StoryboardId = fk_StoryboardId,
                Name = Name,
                Description = Description,
                RenderFailureMessage = RenderFailureMessage,
                DateCreated = DateCreated,
                DateUpdated = DateUpdated,
                DateCompleted = DateCompleted,
                PreviewFilePath = PreviewFilePath,
                CompletedFilePath = CompletedFilePath,
                VidyardVideoId = VidyardVideoId,
                VidyardPlayerId = VidyardPlayerId,
                VidyardPlayerUuid = VidyardPlayerUuid
            };
        }
    }
}
